[BUG] During rolling upgrade, OpenSearch node cannot handling the exception right when security plugin is installed

**Describe the bug**
We found this issue during rolling upgrade test of IM plugin from opendistro cluster to OpenSearch cluster. We have security enabled.

The issue happens when it's a mixed cluster state, for example, 2 opendistro nodes `node-1` and `node-2`, one OpenSearch node `node-3`. 

When you hit `node-3` with request like `curl "https://localhost:9202/nonexist" -u admin:admin --insecure` (simple get request to a non existent index), this request is not returned. And with exception in the log:

<details>
  <summary>Exception, click to see</summary>
  
```
[2021-06-28T16:23:51,104][ERROR][o.o.r.a.RestResponseListener] [node-3] failed to send failure response
java.lang.StringIndexOutOfBoundsException: String index out of range: -3
	at java.lang.String.substring(String.java:1850) ~[?:?]
	at org.opensearch.OpenSearchException.innerToXContent(OpenSearchException.java:351) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.OpenSearchException.toXContent(OpenSearchException.java:339) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.OpenSearchException.generateFailureXContent(OpenSearchException.java:593) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.rest.BytesRestResponse.build(BytesRestResponse.java:162) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:123) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:106) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.rest.action.RestActionListener.onFailure(RestActionListener.java:71) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.action.support.TransportAction$1.onFailure(TransportAction.java:111) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.handleException(TransportMasterNodeAction.java:201) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.TransportService$6.handleException(TransportService.java:662) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.security.transport.SecurityInterceptor$RestoringTransportResponseHandler.handleException(SecurityInterceptor.java:294) [opensearch-security-1.0.0.0-rc1.jar:1.0.0.0-rc1]
	at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1203) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:290) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:239) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundHandler.handleException(InboundHandler.java:288) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:280) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:144) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:713) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:155) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:130) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:95) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
	at org.opensearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:87) [transport-netty4-client-1.0.0-rc1.jar:1.0.0-rc1]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1533) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1282) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1329) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:508) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:447) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
	at java.lang.Thread.run(Thread.java:835) [?:?]
	Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
		at org.opensearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1030) ~[opensearch-x-content-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:122) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:106) ~[opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.rest.action.RestActionListener.onFailure(RestActionListener.java:71) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.action.support.TransportAction$1.onFailure(TransportAction.java:111) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.handleException(TransportMasterNodeAction.java:201) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.TransportService$6.handleException(TransportService.java:662) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.security.transport.SecurityInterceptor$RestoringTransportResponseHandler.handleException(SecurityInterceptor.java:294) [opensearch-security-1.0.0.0-rc1.jar:1.0.0.0-rc1]
		at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1203) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:290) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:239) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundHandler.handleException(InboundHandler.java:288) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:280) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:144) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:713) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:155) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:130) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:95) [opensearch-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:87) [transport-netty4-client-1.0.0-rc1.jar:1.0.0-rc1]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1533) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1282) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1329) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:508) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:447) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at java.lang.Thread.run(Thread.java:835) [?:?]
	Caused by: java.io.IOException: Unclosed object or array found
		at org.opensearch.common.xcontent.json.JsonXContentGenerator.close(JsonXContentGenerator.java:469) ~[opensearch-x-content-1.0.0-rc1.jar:1.0.0-rc1]
		at org.opensearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1028) ~[opensearch-x-content-1.0.0-rc1.jar:1.0.0-rc1]
		... 47 more
	Suppressed: org.opensearch.transport.RemoteTransportException: [node-1][127.0.0.1:9300][indices:admin/get]
	Caused by: org.opensearch.index.IndexNotFoundException: no such index [example1]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.indexNotFoundException(IndexNameExpressionResolver.java:945) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.innerResolve(IndexNameExpressionResolver.java:882) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.resolve(IndexNameExpressionResolver.java:838) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:217) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:197) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:90) ~[?:?]
		at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:52) ~[?:?]
		at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:34) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:138) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:115) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:59) ~[?:?]
		at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:179) ~[?:?]
		at com.amazon.opendistroforelasticsearch.indexmanagement.rollup.actionfilter.FieldCapsFilter.apply(FieldCapsFilter.kt:124) ~[?:?]
		at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:177) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.filter.OpenDistroSecurityFilter.apply0(OpenDistroSecurityFilter.java:311) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.filter.OpenDistroSecurityFilter.apply(OpenDistroSecurityFilter.java:151) ~[?:?]
		at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:177) ~[?:?]
		at com.amazon.opendistro.elasticsearch.performanceanalyzer.action.PerformanceAnalyzerActionFilter.apply(PerformanceAnalyzerActionFilter.java:77) ~[?:?]
		at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:177) ~[?:?]
		at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:155) ~[?:?]
		at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:64) ~[?:?]
		at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:60) ~[?:?]
		at com.amazon.opendistro.elasticsearch.performanceanalyzer.transport.PerformanceAnalyzerTransportRequestHandler.messageReceived(PerformanceAnalyzerTransportRequestHandler.java:49) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.ssl.transport.OpenDistroSecuritySSLRequestHandler.messageReceivedDecorate(OpenDistroSecuritySSLRequestHandler.java:182) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.transport.OpenDistroSecurityRequestHandler.messageReceivedDecorate(OpenDistroSecurityRequestHandler.java:293) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.ssl.transport.OpenDistroSecuritySSLRequestHandler.messageReceived(OpenDistroSecuritySSLRequestHandler.java:142) ~[?:?]
		at com.amazon.opendistroforelasticsearch.security.OpenDistroSecurityPlugin$7$1.messageReceived(OpenDistroSecurityPlugin.java:639) ~[?:?]
		at com.amazon.opendistroforelasticsearch.indexmanagement.rollup.interceptor.RollupInterceptor$interceptHandler$1.messageReceived(RollupInterceptor.kt:124) ~[?:?]
		at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.handleRequest(InboundHandler.java:207) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:107) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:89) ~[?:?]
		at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:700) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82) ~[?:?]
		at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:74) ~[?:?]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1518) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1267) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1314) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) [netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at java.lang.Thread.run(Thread.java:835) [?:?]
```
</details>

At first, we think the problem is with IM plugin or security plugin, so we did different combination tests, results showing in the table. Based on this, the problem is likely in the security plugin.

|                | w/o IM plugin | w/ IM plugin |
|  ---         |     ---      |    ---  |
| w/o security plugin   | not tested     | good   |
| w/ security plugin     | has problem      |  has problem      |

This problem probably can happen for any request that throw exception we believe. **We know it wont happen for `_search` request though**, even search an non-existent index.

**Expected behavior**
The request got back response like 

```
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [example1]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"example1","index":"example1"}],"type":"index_not_found_exception","reason":"no such index [example1]","index_uuid":"_na_","resource.type":"index_or_alias","resource.id":"example1","index":"example1"},"status":404}
```

**Host/Environment (please complete the following information):**
 - OS: Linux
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

The issue is reproducible without the security plugin being installed as well. Please see #905 and run `ExceptionIT.testOpensearchException()` without the fix. Check nodes logs:
```
[2021-06-29T22:43:43,623][ERROR][o.o.r.a.RestResponseListener] [v7.10.2-1] failed to send failure response
java.lang.StringIndexOutOfBoundsException: begin 11, end 8, length 8
	at java.lang.String.checkBoundsBeginEnd(String.java:3756) ~[?:?]
	at java.lang.String.substring(String.java:1902) ~[?:?]
	at java.lang.String.substring(String.java:1875) ~[?:?]
	at org.opensearch.OpenSearchException.innerToXContent(OpenSearchException.java:351) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.OpenSearchException.toXContent(OpenSearchException.java:339) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.OpenSearchException.generateFailureXContent(OpenSearchException.java:593) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.rest.BytesRestResponse.build(BytesRestResponse.java:162) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:123) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:106) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.rest.action.RestActionListener.onFailure(RestActionListener.java:71) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.action.support.TransportAction$1.onFailure(TransportAction.java:111) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.handleException(TransportMasterNodeAction.java:201) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.TransportService$6.handleException(TransportService.java:662) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1203) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:290) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:239) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.handleException(InboundHandler.java:288) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:280) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:144) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:713) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:155) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:130) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:95) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at org.opensearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:87) [transport-netty4-client-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
	at java.lang.Thread.run(Thread.java:832) [?:?]
	Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
		at org.opensearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1030) ~[opensearch-x-content-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:122) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:106) ~[opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.rest.action.RestActionListener.onFailure(RestActionListener.java:71) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.action.support.TransportAction$1.onFailure(TransportAction.java:111) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$1.handleException(TransportMasterNodeAction.java:201) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.TransportService$6.handleException(TransportService.java:662) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1203) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundHandler.lambda$handleException$3(InboundHandler.java:290) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.common.util.concurrent.OpenSearchExecutors$DirectExecutorService.execute(OpenSearchExecutors.java:239) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundHandler.handleException(InboundHandler.java:288) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundHandler.handlerResponseError(InboundHandler.java:280) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundHandler.messageReceived(InboundHandler.java:144) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundHandler.inboundMessage(InboundHandler.java:102) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.TcpTransport.inboundMessage(TcpTransport.java:713) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:155) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:130) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:95) [opensearch-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:87) [transport-netty4-client-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at java.lang.Thread.run(Thread.java:832) [?:?]
	Caused by: java.io.IOException: Unclosed object or array found
		at org.opensearch.common.xcontent.json.JsonXContentGenerator.close(JsonXContentGenerator.java:469) ~[opensearch-x-content-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		at org.opensearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1028) ~[opensearch-x-content-1.0.0-SNAPSHOT.jar:1.0.0-SNAPSHOT]
		... 37 more
	Suppressed: org.opensearch.transport.RemoteTransportException: [v7.10.2-2][127.0.0.1:65408][indices:admin/get]
	Caused by: org.opensearch.index.IndexNotFoundException: no such index [no_such_index]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.indexNotFoundException(IndexNameExpressionResolver.java:945) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.innerResolve(IndexNameExpressionResolver.java:882) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.resolve(IndexNameExpressionResolver.java:838) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:217) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:197) ~[?:?]
		at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:90) ~[?:?]
		at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:52) ~[?:?]
		at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:34) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:138) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:115) ~[?:?]
		at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:59) ~[?:?]
		at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:179) ~[?:?]
		at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:155) ~[?:?]
		at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:64) ~[?:?]
		at org.elasticsearch.action.support.HandledTransportAction$TransportHandler.messageReceived(HandledTransportAction.java:60) ~[?:?]
		at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:72) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.handleRequest(InboundHandler.java:207) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:107) ~[?:?]
		at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:89) ~[?:?]
		at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:700) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:142) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:117) ~[?:?]
		at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:82) ~[?:?]
		at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:74) ~[?:?]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[netty-codec-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.59.Final.jar:4.1.59.Final]
		at java.lang.Thread.run(Thread.java:832) [?:?]
```

--- a/qa/mixed-cluster/src/test/java/org/opensearch/backwards/ExceptionIT.java
+++ b/qa/mixed-cluster/src/test/java/org/opensearch/backwards/ExceptionIT.java
@@ -1,0 +1,61 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.backwards;
+
+import org.apache.http.util.EntityUtils;
+import org.opensearch.Version;
+import org.opensearch.client.Node;
+import org.opensearch.client.Request;
+import org.opensearch.client.Response;
+import org.opensearch.client.ResponseException;
+import org.opensearch.test.rest.OpenSearchRestTestCase;
+import org.opensearch.test.rest.yaml.ObjectPath;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+import static org.apache.http.HttpStatus.SC_NOT_FOUND;
+
+public class ExceptionIT extends OpenSearchRestTestCase {
+    public void testOpensearchException() throws Exception {
+        logClusterNodes();
+
+        Request request = new Request("GET", "/no_such_index");
+
+        for (Node node : client().getNodes()) {
+            try {
+                client().setNodes(Collections.singletonList(node));
+                logger.info("node: {}", node.getHost());
+                client().performRequest(request);
+                fail();
+            } catch (ResponseException e) {
+                logger.debug(e.getMessage());
+                Response response = e.getResponse();
+                assertEquals(SC_NOT_FOUND, response.getStatusLine().getStatusCode());
+                assertEquals("no_such_index", ObjectPath.createFromResponse(response).evaluate("error.index"));
+            }
+        }
+    }
+
+    private void logClusterNodes() throws IOException {
+        ObjectPath objectPath = ObjectPath.createFromResponse(client().performRequest(new Request("GET", "_nodes")));
+        Map<String, ?> nodes = objectPath.evaluate("nodes");
+        String master = EntityUtils.toString(client().performRequest(new Request("GET", "_cat/master?h=id")).getEntity()).trim();
+        logger.info("cluster discovered: master id='{}'", master);
+        for (String id : nodes.keySet()) {
+            logger.info("{}: id='{}', name='{}', version={}",
+                objectPath.evaluate("nodes." + id + ".http.publish_address"),
+                id,
+                objectPath.evaluate("nodes." + id + ".name"),
+                Version.fromString(objectPath.evaluate("nodes." + id + ".version"))
+            );
+        }
+    }
+}

--- a/server/src/main/java/org/opensearch/OpenSearchException.java
+++ b/server/src/main/java/org/opensearch/OpenSearchException.java
@@ -61,6 +61,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import static java.util.Collections.emptyMap;
@@ -108,6 +109,10 @@ public class OpenSearchException extends RuntimeException implements ToXContentF
 
     private static final Map<Integer, CheckedFunction<StreamInput, ? extends OpenSearchException, IOException>> ID_TO_SUPPLIER;
     private static final Map<Class<? extends OpenSearchException>, OpenSearchExceptionHandle> CLASS_TO_OPENSEARCH_EXCEPTION_HANDLE;
+
+    private static final Pattern OS_METADATA = Pattern.compile("^opensearch\\.");
+    private static final Pattern ES_METADATA = Pattern.compile("^es\\.");
+
     private final Map<String, List<String>> metadata = new HashMap<>();
     private final Map<String, List<String>> headers = new HashMap<>();
 
@@ -150,7 +155,16 @@ public class OpenSearchException extends RuntimeException implements ToXContentF
         super(in.readOptionalString(), in.readException());
         readStackTrace(this, in);
         headers.putAll(in.readMapOfLists(StreamInput::readString, StreamInput::readString));
-        metadata.putAll(in.readMapOfLists(StreamInput::readString, StreamInput::readString));
+        metadata.putAll(in.readMapOfLists(OpenSearchException::readAndReplace, StreamInput::readString));
+    }
+
+    private static String readAndReplace(StreamInput in) throws IOException {
+        String str = in.readString();
+        return in.getVersion().onOrBefore(LegacyESVersion.V_7_10_2) ? ES_METADATA.matcher(str).replaceFirst("opensearch.") : str;
+    }
+
+    private static void replaceAndWrite(StreamOutput out, String str) throws IOException {
+        out.writeString(out.getVersion().onOrBefore(LegacyESVersion.V_7_10_2) ? OS_METADATA.matcher(str).replaceFirst("es.") : str);
     }
 
     /**
@@ -293,7 +307,7 @@ public class OpenSearchException extends RuntimeException implements ToXContentF
         out.writeException(this.getCause());
         writeStackTraces(this, out, StreamOutput::writeException);
         out.writeMapOfLists(headers, StreamOutput::writeString, StreamOutput::writeString);
-        out.writeMapOfLists(metadata, StreamOutput::writeString, StreamOutput::writeString);
+        out.writeMapOfLists(metadata, OpenSearchException::replaceAndWrite, StreamOutput::writeString);
     }
 
     public static OpenSearchException readException(StreamInput input, int id) throws IOException {

